runtime/bindings/python/hal.cc

```diff
@@ -286,13 +308,30 @@ HalDevice HalDriver::CreateDevice(iree_hal_device_id_t device_id) {
     throw std::invalid_argument(std::move(msg));
   }
 
+  std::vector<iree_string_pair_t> params;
   iree_hal_device_t* device;
-  CheckApiStatus(iree_hal_driver_create_device(
-                     raw_ptr(), device_id, iree_allocator_system(), &device),
+  CheckApiStatus(iree_hal_driver_create_device_by_id(
+                     raw_ptr(), device_id, params.size(), &params.front(),
+                     iree_allocator_system(), &device),
                  "Error creating default device");
   return HalDevice::StealFromRawPtr(device);
 }
 
+HalDevice HalDriver::CreateDeviceByURI(std::string& device_uri) {
+  iree_hal_device_t* device;
+  iree_string_view_t device_uri_sv{device_uri.data(), device_uri.size()};
+  iree_status_t status = iree_hal_driver_create_device_by_uri(
+      raw_ptr(), device_uri_sv, iree_allocator_system(), &device);
+  if (iree_status_is_not_found(status) ||
```

Comment:
this will leak the status - is that ok? (or are you throwing to kill the process?)
if you want to discard the status then you can use `iree_status_ignore(status)`

the status does include the the device URI already in addition to other info so it'd probably be worth doing what you do in CheckApiStatus to get the message out instead:
https://github.com/google/iree/blob/42e3eb6aad6418775c0f719aa606720d38d5910e/runtime/src/iree/hal/driver.c#L164-L165

Reply:
Realized that CheckApiStatus was already translating to exceptions but was missing this case. Just added and used that. Also changed the UNIMPLEMENTED in local drivers to NOT_FOUND, which is actually what it should be signalling -- and that removes the special case.